test_runner/batch_others/test_wal_acceptor.py

```diff
@@ -0,0 +1,199 @@
+# succeed and data is written
+def test_normal_work(zenith_cli, pageserver, postgres, wa_factory):
+    zenith_cli.run(["branch", "test_wal_acceptors_normal_work", "empty"])
+    wa_factory.start_n_new(3)
+    pg = postgres.create_start('test_wal_acceptors_normal_work',
+                               wal_acceptors=wa_factory.get_connstrs())
+
+    pg_conn = psycopg2.connect(pg.connstr())
+    # do commit after each statement as waiting for acceptors happens there
+    pg_conn.autocommit = True
+
+    cur = pg_conn.cursor()
+    cur.execute('CREATE TABLE t(key int primary key, value text)')
+    cur.execute("INSERT INTO t SELECT generate_series(1,100000), 'payload'")
+    cur.execute('SELECT sum(key) FROM t')
+    assert cur.fetchone() == (5000050000, )
+
+
+# Run page server and multiple acceptors, and multiple compute nodes running
+# against different timelines.
+def test_many_timelines(zenith_cli, pageserver, postgres, wa_factory):
+    n_timelines = 2
+
+    wa_factory.start_n_new(3)
+
+    branches = ["test_wal_acceptors_many_timelines_{}".format(tlin) for tlin in range(n_timelines)]
+
+    # start postgres on each timeline
+    pgs = []
+    for branch in branches:
+        zenith_cli.run(["branch", branch, "empty"])
+        pgs.append(postgres.create_start(branch, wal_acceptors=wa_factory.get_connstrs()))
+
+    # create schema
+    for pg in pgs:
+        pg.safe_psql("CREATE TABLE t(key int primary key, value text)")
+
+    # Populate data
+    for pg in pgs:
+        pg.safe_psql("INSERT INTO t SELECT generate_series(1,100000), 'payload'")
+
+    # Check data
+    for pg in pgs:
+        res = pg.safe_psql("SELECT sum(key) FROM t")
+        assert res[0] == (5000050000, )
+
+
+# Execute insert n_inserts times, with fault_probability chance of getting a wal
+# acceptor down or up along the way. 2 of 3 are always alive, so the work keeps
+# going.
+def test_restarts(zenith_cli, pageserver, postgres, wa_factory):
+    fault_probability = 0.01
+    n_inserts = 1000
+    n_acceptors = 3
+
+    wa_factory.start_n_new(n_acceptors)
+
+    zenith_cli.run(["branch", "test_wal_acceptors_restarts", "empty"])
+    pg = postgres.create_start('test_wal_acceptors_restarts',
+                               wal_acceptors=wa_factory.get_connstrs())
+
+    pg_conn = psycopg2.connect(pg.connstr())
+    # do commit after each statement as waiting for acceptors happens there
+    pg_conn.autocommit = True
+
+    cur = pg_conn.cursor()
+
+    failed_node = None
+    cur.execute('CREATE TABLE t(key int primary key, value text)')
+    for i in range(n_inserts):
+        cur.execute("INSERT INTO t values (%s, 'payload');", (i + 1, ))
+
+        if random.random() <= fault_probability:
+            if failed_node is None:
+                failed_node = wa_factory.instances[random.randrange(0, n_acceptors)]
+                failed_node.stop()
+            else:
+                failed_node.start()
+                failed_node = None
```

Comment:
As has been discussed in private: overall, I'd like to see more comments outlining test design decisions and goals.

Reply:
added the comment

Reply:
Tried to do it a bit better.